Mathlib/Analysis/Normed/Ring/IsPowMulFaithful.lean

```diff
@@ -0,0 +1,120 @@
+
+/-- If `f : α →+* β` is bounded with respect to a ring seminorm `nα` on `α` and a
+  power-multiplicative function `nβ : β → ℝ`, then `∀ x : α, nβ (f x) ≤ nα x`. -/
+theorem contraction_of_isPowMul_of_boundedWrt {F : Type*} {α : outParam (Type*)} [Ring α]
+    [FunLike F α ℝ] [RingSeminormClass F α ℝ] {β : Type*} [Ring β] (nα : F) {nβ : β → ℝ}
+    (hβ : IsPowMul nβ) {f : α →+* β} (hf : f.IsBoundedWrt nα nβ) (x : α) : nβ (f x) ≤ nα x := by
+  obtain ⟨C, hC0, hC⟩ := hf
+  have hlim : Tendsto (fun n : ℕ => C ^ (1 / (n : ℝ)) * nα x) atTop (𝓝 (nα x)) := by
+    nth_rewrite 2 [← one_mul (nα x)]
+    exact ((rpow_zero C ▸ ContinuousAt.tendsto (continuousAt_const_rpow (ne_of_gt hC0))).comp
+      (tendsto_const_div_atTop_nhds_zero_nat 1)).mul tendsto_const_nhds
+  apply ge_of_tendsto hlim
+  simp only [eventually_atTop, ge_iff_le]
+  use 1
+  intro n hn
+  have h : (C ^ (1 / n : ℝ)) ^ n = C := by
+    have hn0 : (n : ℝ) ≠ 0 := Nat.cast_ne_zero.mpr (ne_of_gt hn)
+    rw [← rpow_natCast, ← rpow_mul (le_of_lt hC0), one_div, inv_mul_cancel hn0, rpow_one]
+  apply le_of_pow_le_pow_left (ne_of_gt hn)
+    (mul_nonneg (rpow_nonneg (le_of_lt hC0) _) (apply_nonneg _ _))
+  · rw [mul_pow, h, ← hβ _ hn, ← RingHom.map_pow]
+    apply le_trans (hC (x ^ n))
+    rw [mul_le_mul_left hC0]
+    exact map_pow_le_pow _ _ (Nat.one_le_iff_ne_zero.mp hn)
+
+/-- Given a bounded `f : α →+* β` between seminormed rings, is the seminorm on `β` is
+  power-multiplicative, then `f` is a contraction. -/
+theorem contraction_of_isPowMul {α β : Type*} [SeminormedRing α] [SeminormedRing β]
+    (hβ : IsPowMul (norm : β → ℝ)) {f : α →+* β} (hf : f.IsBounded) (x : α) : norm (f x) ≤ norm x :=
+  contraction_of_isPowMul_of_boundedWrt (SeminormedRing.toRingSeminorm α) hβ hf x
+
+/-- Given two power-multiplicative ring seminorms `f, g` on `α`, if `f` is bounded by a positive
+  multiple of `g` and viceversa, then `f = g`. -/
+theorem eq_seminorms {F : Type*} {α : outParam (Type*)} [Ring α] [FunLike F α ℝ]
+    [RingSeminormClass F α ℝ] {f g : F} (hfpm : IsPowMul f) (hgpm : IsPowMul g)
+    (hfg : ∃ (r : ℝ) (_ : 0 < r), ∀ a : α, f a ≤ r * g a)
+    (hgf : ∃ (r : ℝ) (_ : 0 < r), ∀ a : α, g a ≤ r * f a) : f = g := by
+  obtain ⟨r, hr0, hr⟩ := hfg
+  obtain ⟨s, hs0, hs⟩ := hgf
+  have hle : RingHom.IsBoundedWrt f g (RingHom.id _) := ⟨s, hs0, hs⟩
+  have hge : RingHom.IsBoundedWrt g f (RingHom.id _) := ⟨r, hr0, hr⟩
+  rw [← Function.Injective.eq_iff DFunLike.coe_injective']
+  ext x
+  exact le_antisymm (contraction_of_isPowMul_of_boundedWrt g hfpm hge x)
+    (contraction_of_isPowMul_of_boundedWrt f hgpm hle x)
+
+variable {R S : Type _} [NormedCommRing R] [CommRing S] [Algebra R S]
+
+/-- The restriction of a power-multiplicative function to a subalgebra is power-multiplicative. -/
+theorem IsPowMul.restriction (A : Subalgebra R S) {f : S → ℝ} (hf_pm : IsPowMul f) :
+    IsPowMul fun x : A => f x.val := fun x n hn => by
+  simpa [SubsemiringClass.coe_pow] using hf_pm (↑x) hn
```

Comment:
This should probably go somewhere earlier as well...

Reply:
I moved it to `Analysis.Normed.Field.Basic`.